### DISCUSSION TITLE
[help] tar_render within tar_map for parametrized mapped targets

### DISCUSSION THREAD

**Discussion by CorradoLanera** [ropensci/tarchetypes#181] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I have a `tar_map` that generates multiple versions of a dataset and a report to compile on each of them, I would need also to pass the `tar_map`'s values to the report to know which set of parameters compose the report (but this is not an issue). The point is that all of this works fine, but all the reports will have the same name, so they overwrite each other, and creating the output name by pasting a base name with the `tar_map`'s values leads to an error. The strange thing to me is that the values have been found to pass them as parameters of the report, but not if used to compose the name.
Note: Using `tar_render_rep` works fine, but I completely lost the advantage of static branching and cannot connect output names with the parameters used.

```` r
# Works but overwrite results
targets::tar_dir({
  targets::tar_script({
    c("---
title: report
output: md_document
params:
  par: 4
  db: mtcars
---

```{r, echo = FALSE}
cat('par:', params$par)
cat('db:', dim(params$db))
cat('val:', params$db$val)
```
  ") |>
      writeLines("report.Rmd")

    tarchetypes::tar_map(
      values = list(.par = 1:2),
      targets::tar_target(val, data.frame(val = .par)),
      tarchetypes::tar_render(
        report,
        "report.Rmd",
        # output = paste0("report_", .par, ".html"),
        params = list(par = .par, db = val)
      )
    )
  })

  targets::tar_make()
  cat(list.files(), "\n")
  readLines("report.md") |> paste0(collapse = "\n") |> cat()
})
````

    #> ▶ dispatched target val_1
    #> ● completed target val_1 [0 seconds]
    #> ▶ dispatched target val_2
    #> ● completed target val_2 [0 seconds]
    #> ▶ dispatched target report_1
    #> ● completed target report_1 [0.375 seconds]
    #> ▶ dispatched target report_2
    #> ● completed target report_2 [0.125 seconds]
    #> ▶ ended pipeline [0.562 seconds]
    #> _targets _targets.R report.md report.Rmd 
    #>     ## par: 2
    #> 
    #>     ## db: 1 1
    #> 
    #>     ## val: 2

```` r


# Does not find `.par` to compose the output name
targets::tar_dir({
  targets::tar_script({
    c("---
title: report
output: md_document
params:
  par: 4
  db: mtcars
---

```{r, echo = FALSE}
cat('par:', params$par)
cat('db:', dim(params$db))
cat('val:', params$db$val)
```
  ") |>
      writeLines("report.Rmd")

    tarchetypes::tar_map(
      values = list(.par = 1:2),
      targets::tar_target(val, data.frame(val = .par)),
      tarchetypes::tar_render(
        report,
        "report.Rmd",
        output = paste0("report_", .par, ".html"),
        params = list(par = .par, db = val)
      )
    )
  })

  targets::tar_make()
  cat(list.files(), "\n")
  readLines("report.md") |> paste0(collapse = "\n") |> cat()
})
````

    #> Error:
    #> ! Error running targets::tar_make()
    #> Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
    #> Debugging guide: https://books.ropensci.org/targets/debugging.html
    #> How to ask for help: https://books.ropensci.org/targets/help.html
    #> Last error message:
    #>     object '.par' not found
    #> Last error traceback:
    #>     base::tryCatch(base::withCallingHandlers({ NULL base::saveRDS(base::do.c...
    #>     tryCatchList(expr, classes, parentenv, handlers)
    #>     tryCatchOne(tryCatchList(expr, names[-nh], parentenv, handlers[-nh]), na...
    #>     doTryCatch(return(expr), name, parentenv, handler)
    #>     tryCatchList(expr, names[-nh], parentenv, handlers[-nh])
    #>     tryCatchOne(expr, names, parentenv, handlers[[1L]])
    #>     doTryCatch(return(expr), name, parentenv, handler)
    #>     base::withCallingHandlers({ NULL base::saveRDS(base::do.call(base::do.ca...
    #>     base::saveRDS(base::do.call(base::do.call, base::c(base::readRDS("C:\\Us...
    #>     base::do.call(base::do.call, base::c(base::readRDS("C:\\Users\\corra\\Ap...
    #>     (function (what, args, quote = FALSE, envir = parent.frame()) { if (!is....
    #>     (function (targets_function, targets_arguments, options, envir = NULL, s...
    #>     tryCatch(out <- withCallingHandlers(targets::tar_callr_inner_try(targets...
    #>     tryCatchList(expr, classes, parentenv, handlers)
    #>     tryCatchOne(expr, names, parentenv, handlers[[1L]])
    #>     doTryCatch(return(expr), name, parentenv, handler)
    #>     withCallingHandlers(targets::tar_callr_inner_try(targets_function = targ...
    #>     targets::tar_callr_inner_try(targets_function = targets_function, target...
    #>     eval(parse(file = script, keep.source = TRUE), envir = envir)
    #>     eval(parse(file = script, keep.source = TRUE), envir = envir)
    #>     tarchetypes::tar_map(values = list(.par = 1:2), targets::tar_target(val,...
    #>     unlist(list(...), recursive = TRUE) %|||% list()
    #>     unlist(list(...), recursive = TRUE)
    #>     tarchetypes::tar_render(report, "report.Rmd", output = paste0("report_",...
    #>     targets::tar_assert_chr(output_file %|||% "x")
    #>     output_file %|||% "x"
    #>     paste0("report_", .par, ".html")

```` r



## Works with dynamic branching
targets::tar_dir({
  targets::tar_script({
    c("---
title: report
output: md_document
params:
  par: 4
  db: mtcars
---

```{r, echo = FALSE}
cat('par:', params$par)
cat('db:', dim(params$db))
cat('val:', params$db$val)
```
  ") |>
      writeLines("report.Rmd")

    tarchetypes::tar_map(
      values = list(.par = 1:2),
      targets::tar_target(val, data.frame(val = .par)),
      tarchetypes::tar_render_rep(
        report,
        "report.Rmd",
        params = data.frame(par = .par, db = I(val))
      )
    )
  })

  targets::tar_make()
  cat(list.files(), "\n")
  list.files(pattern = "\\.md$") |>
    purrr::map_chr(\(x) readLines(x) |> paste0(collapse = "\n")) |>
    paste0(collapse = "\n\n\n") |>
    cat()
})
````

    #> ▶ dispatched target val_1
    #> ● completed target val_1 [0 seconds]
    #> ▶ dispatched target val_2
    #> ● completed target val_2 [0 seconds]
    #> ▶ dispatched target report_params_1
    #> ● completed target report_params_1 [0 seconds]
    #> ▶ dispatched target report_params_2
    #> ● completed target report_params_2 [0 seconds]
    #> ▶ dispatched branch report_1_5ca6935ef3cb82e6
    #> ● completed branch report_1_5ca6935ef3cb82e6 [0.407 seconds]
    #> ● completed pattern report_1
    #> ▶ dispatched branch report_2_beb7222a396608f4
    #> ● completed branch report_2_beb7222a396608f4 [0.14 seconds]
    #> ● completed pattern report_2
    #> ▶ ended pipeline [0.656 seconds]
    #> _targets _targets.R report.Rmd report_6375abe78e94b8c1.md report_f8b02cf4f188fb41.md 
    #>     ## par: 2
    #> 
    #>     ## db: 1 1
    #> 
    #>     ## val: 2
    #> 
    #> 
    #>     ## par: 1
    #> 
    #>     ## db: 1 1
    #> 
    #>     ## val: 1

<sup>Created on 2024-07-05 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.4.1 (2024-06-14 ucrt)
#>  os       Windows 11 x64 (build 26120)
#>  system   x86_64, mingw32
#>  ui       RTerm
#>  language (EN)
#>  collate  English_United States.utf8
#>  ctype    English_United States.utf8
#>  tz       Europe/Berlin
#>  date     2024-07-05
#>  pandoc   3.1.11 @ C:/Program Files/RStudio/resources/app/bin/quarto/bin/tools/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  ! package     * version date (UTC) lib source
#>  P backports     1.5.0   2024-05-23 [?] RSPM
#>  P base64url     1.4     2018-05-14 [?] CRAN (R 4.4.0)
#>  P callr         3.7.6   2024-03-25 [?] RSPM (R 4.4.0)
#>  P cli           3.6.3   2024-06-21 [?] RSPM
#>  P codetools     0.2-20  2024-03-31 [3] CRAN (R 4.4.1)
#>  P data.table    1.15.4  2024-03-30 [?] CRAN (R 4.4.0)
#>  P digest        0.6.36  2024-06-23 [?] CRAN (R 4.4.1)
#>  P evaluate      0.24.0  2024-06-10 [?] RSPM
#>  P fansi         1.0.6   2023-12-08 [?] RSPM (R 4.4.0)
#>  P fastmap       1.2.0   2024-05-15 [?] RSPM
#>  P fs            1.6.4   2024-04-25 [?] RSPM (R 4.4.0)
#>  P glue          1.7.0   2024-01-09 [?] RSPM (R 4.4.0)
#>  P htmltools     0.5.8.1 2024-04-04 [?] RSPM (R 4.4.0)
#>  P igraph        2.0.3   2024-03-13 [?] CRAN (R 4.4.0)
#>  P knitr         1.47    2024-05-29 [?] CRAN (R 4.4.0)
#>  P lifecycle     1.0.4   2023-11-07 [?] RSPM (R 4.4.0)
#>  P magrittr      2.0.3   2022-03-30 [?] RSPM (R 4.4.0)
#>  P pillar        1.9.0   2023-03-22 [?] RSPM (R 4.4.0)
#>  P pkgconfig     2.0.3   2019-09-22 [?] RSPM (R 4.4.0)
#>  P processx      3.8.4   2024-03-16 [?] RSPM (R 4.4.0)
#>  P ps            1.7.7   2024-07-02 [?] CRAN (R 4.4.1)
#>  P purrr         1.0.2   2023-08-10 [?] RSPM (R 4.4.0)
#>  P R6            2.5.1   2021-08-19 [?] RSPM (R 4.4.0)
#>  P reprex        2.1.0   2024-01-11 [?] CRAN (R 4.4.0)
#>  P rlang         1.1.4   2024-06-04 [?] CRAN (R 4.4.0)
#>  P rmarkdown     2.27    2024-05-17 [?] RSPM
#>  P rstudioapi    0.16.0  2024-03-24 [?] RSPM (R 4.4.0)
#>  P secretbase    1.0.0   2024-06-16 [?] CRAN (R 4.4.1)
#>  P sessioninfo   1.2.2   2021-12-06 [?] RSPM (R 4.4.0)
#>  P targets       1.7.1   2024-06-20 [?] CRAN (R 4.4.1)
#>  P tibble        3.2.1   2023-03-20 [?] RSPM (R 4.4.0)
#>  P tidyselect    1.2.1   2024-03-11 [?] RSPM (R 4.4.0)
#>  P utf8          1.2.4   2023-10-22 [?] RSPM (R 4.4.0)
#>  P vctrs         0.6.5   2023-12-01 [?] RSPM (R 4.4.0)
#>  P withr         3.0.0   2024-01-16 [?] RSPM (R 4.4.0)
#>  P xfun          0.45    2024-06-16 [?] RSPM
#>  P yaml          2.3.8   2023-12-11 [?] RSPM (R 4.4.0)
#> 
#>  [1] C:/Users/corra/AppData/Local/R/cache/R/renv/library/anvur.accademies.gpt-ebd9193f/windows/R-4.4/x86_64-w64-mingw32
#>  [2] C:/Users/corra/AppData/Local/R/cache/R/renv/sandbox/windows/R-4.4/x86_64-w64-mingw32/e0da0d43
#>  [3] C:/Program Files/R/R-4.4.1/library
#> 
#>  P ── Loaded and on-disk path mismatch.
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

**Comment by ginolhac** [comment]
Hello,
maybe this helps: `tar_quarto_rep()` using static branching. https://github.com/ropensci/targets/discussions/1297#discussioncomment-9728095

**Reply by wlandau** [reply]
That's a great suggestion. Alternatively, `tar_eval()` is a more flexible kind of static branching that can (unlike `tar_map()`) branch over calls to `tar_render()`: https://books.ropensci.org/targets/static.html#metaprogramming

**Reply by CorradoLanera** [reply, marked as answer]
@ginolhac , your suggestion was fantastic and permitted me to solve the issue, but not directly that way. 
I have to create a little more complex example to explain way, I'll do that in the next few days (after finish the project). For the moment: big thank you!

basically the main points are:
- we need a dataframe of parameters (not a list)
- we need a dataframe of parameters with a column containing dataframes (the parametrized data)
- we need a dataframe of parameters containing only parameters columns. plus report output names
- `name`, and `value` are the default colnames for `separate_wider_*`
- we need to change the targets' parameters names into the report's parameters names

and, on top of everything, it is your suggestion: `tar_combine`!


## reprex solution

```` r
# My solution
targets::tar_dir({
  targets::tar_script({
    c("---
title: report
output: md_document
params:
  par: 4
  db: mtcars
---

```{r, echo = FALSE}
cat('par:', params$par)
# data are passed inside a list, so params$db[[1]] to extract them
cat('db dims:', dim(params$db[[1]]))
cat('val:', params$db[[1]]$val)
```
  ") |>
      writeLines("report.Rmd")

    mapped <- tarchetypes::tar_map(
      values = list(.par = 1:2),
      targets::tar_target(val, data.frame(val = .par))
    )

    list(
      mapped,

      tarchetypes::tar_combine(
        combi,
        mapped[["val"]],
        # we have dataframes as parameters so we need some black magic
        # here to have a dataframe with a column containing dataframes
        command = tibble::enframe(list(!!!.x)) |>
          tidyr::separate_wider_position(  # extract parameters
            name,
            widths = c(par = 3, 1, val = 1),
            cols_remove = FALSE # `name` is needed for filename
          ) |>
          dplyr::mutate( # set file names
            output_file = paste0(
              stringr::str_remove_all(lubridate::today(), "\\D+"),
              "_repoort_", name, ".md"
            )
          ) |>
          dplyr::select(-name, -val) |> # removed because not params!
          dplyr::rename(db = value) # param name inside report
      ),

      tarchetypes::tar_render_rep(
        report,
        here::here("report.Rmd"),
        params = combi
      )

    )
  })

  targets::tar_make()
  cat(list.files(), "\n")
  reports <- stringr::str_c(
    stringr::str_remove_all(lubridate::today(), "\\D+"),
    "_repoort_val_", 1:2, ".md"
  )
  readLines(reports[[1]]) |> paste0(collapse = "\n") |> cat()
  cat("\n\n\n")
  readLines(reports[[2]]) |> paste0(collapse = "\n") |> cat()
})
````

    #> ▶ dispatched target val_1
    #> ● completed target val_1 [0 seconds]
    #> ▶ dispatched target val_2
    #> ● completed target val_2 [0 seconds]
    #> ▶ dispatched target combi
    #> ● completed target combi [0.282 seconds]
    #> ▶ dispatched target report_params
    #> ● completed target report_params [0 seconds]
    #> ▶ dispatched branch report_1fcd0558e4bb4425
    #> ● completed branch report_1fcd0558e4bb4425 [0.375 seconds]
    #> ▶ dispatched branch report_d985bc7d867306e3
    #> ● completed branch report_d985bc7d867306e3 [0.14 seconds]
    #> ● completed pattern report
    #> ▶ ended pipeline [0.922 seconds]
    #> _targets _targets.R 20240711_repoort_val_1.md 20240711_repoort_val_2.md report.Rmd 
    #>     ## par: val
    #> 
    #>     ## db dims: 1 1
    #> 
    #>     ## val: 1
    #> 
    #> 
    #>     ## par: val
    #> 
    #>     ## db dims: 1 1
    #> 
    #>     ## val: 2

<sup>Created on 2024-07-11 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by ginolhac** [reply]
very nice, well done! Looking forward to your next comment when you are done.
Yes indeed, the `data.frame` trick is helping a lot, also its column names which must fit `params`.
static branching is intimidating and it took me some time (and also @wlandau answers here) to have like nested static branching working. But now I embraced `tar_combine()` inside nested, or outside and it is a bliss
